### DISCUSSION TITLE
Unable to Access Postgres with DistributedApplicationTestingBuilder

### DISCUSSION THREAD

**Discussion by batesjernigan-aiwyn** [microsoft/aspire#4657] [Q&A]
## Summary
I can't access Postgres for seeding / data setup in integration tests for Tests using Aspire

## Detail
I'm starting to write integration tests in a new aspire project. I'm trying to add some test data to the project before I run the test code and make my assertions.

The ConnectionStrings I have setup in my appsettings.json files are placeholders and I use aspire to build them at Runtime. When I try and access my ConnectionStrings, I get the following error

```csharp
var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AppHost>();
var pg = appHost.Resources.OfType<PostgresServerResource>().First(r => r.Name == "PostgresServer"); // this returns the correct Resource.

Console.WriteLine(pg.ConnectionStringExpression.ValueExpression)
// => Host={PostgresServer.bindings.tcp.host};Port={PostgresServer.bindings.tcp.port};Username=postgres;Password={PostgresServer-password.value}

var connString = await pg.GetConnectionStringAsync(); // Error

/**
System.InvalidOperationException: The endpoint `tcp` is not allocated for the resource `PostgresServer`.
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_AllocatedEndpoint() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 107
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_Host() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 89
   at Aspire.Hosting.ApplicationModel.EndpointReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 177
   at Aspire.Hosting.ApplicationModel.ReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/ReferenceExpression.cs:line 66
**/
```
I see that there's a slightly different way to write tests that uses [DistrubutedApplication directly](https://github.com/dotnet/aspire/blob/6521a859a0f51eec17fca873a313ba0d6bb45d6a/tests/Aspire.Hosting.Tests/Postgres/AddPostgresTests.cs#L144-L155) instead of `DistributedApplicationTestingBuilder`. This looks like it would work for what I need, but wasn't sure if there's some unseen issue with using that directly.

I may be missing a very core / fundamental setup thing about dotnet as I'm still learning this stack and aspire in general. 

## Question
1. Does my setup look correct? 
2. Is there some other recommended way to add data to the DB for an integration test?

FWIW Here's my test setup
```csharp
public class OidcClientProviderTest
{
    [Fact]
    public async Task ShouldLoginAndGetAccessToken()
    {
        // Arrange
        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AppHost>();
        var pg = appHost.Resources.OfType<PostgresServerResource>().First(r => r.Name == "PostgresServer"); // this returns the correct Resource.

        Console.WriteLine(pg.ConnectionStringExpression.ValueExpression)
        // => Host={PostgresServer.bindings.tcp.host};Port={PostgresServer.bindings.tcp.port};Username=postgres;Password={PostgresServer-password.value}
        var connString = await pg.GetConnectionStringAsync(); // Error
    }
}
```

**Comment by davidfowl** [comment]
You have to Start the application before you can get the connection string.

**Reply by batesjernigan-aiwyn** [reply]
Thank you for the reply. That makes sense, but unfortunately I'm still having issues. I tried to find the ConnectionString due to the below error, here's my actual test and what I'm trying to do in my test. I want to start my AppHost app, then I want to setup the DB with an Application record (and probably a User record) so that I can test that login is working correctly with those seed values.

When I try to write this test.
```csharp
public class OidcClientProviderTest
{
    [Fact]
    public async Task ShouldLoginAndGetAccessToken()
    {
        // Arrange
        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AppHost>();
        appHost.Services.AddDbContext<DataContext>(optionsBuilder =>
        {
            optionsBuilder.UseNpgsql(sql =>
            {
                sql.CommandTimeout(30);
                sql.MigrationsAssembly(typeof(DataContext).Assembly.GetName().Name);
                sql.EnableRetryOnFailure(2, TimeSpan.FromSeconds(2), null);
                sql.UseQuerySplittingBehavior(QuerySplittingBehavior.SingleQuery);
            });
            optionsBuilder.ConfigureWarnings(w =>
            {
                w.Throw(RelationalEventId.MultipleCollectionIncludeWarning);
            });
            optionsBuilder.EnableDetailedErrors();
            optionsBuilder.UseExceptionProcessor();
            optionsBuilder.EnableSensitiveDataLogging();
            optionsBuilder.UseSnakeCaseNamingConvention();
        });
        appHost.Services.AddOpenIddict()
            .AddCore(options =>
            {
                options.UseEntityFrameworkCore(efOptions =>
                {
                    efOptions.UseDbContext<DataContext>()
                        .ReplaceDefaultEntities<Application, Authorization, Scope, Token, Guid>();
                });
            });

        await using var app = await appHost.BuildAsync();
        await app.StartAsync();

        // Test Application
        var testApp = new OpenIddictApplicationDescriptor
        {
            ClientId = "TestClient",
            ClientSecret = "TestClientSecret",
            Requirements = { OpenIddictConstants.Requirements.Features.ProofKeyForCodeExchange },
            DisplayName = "Administration",
            RedirectUris = {new Uri("https://localhost:7264/admin")},
            PostLogoutRedirectUris ={new Uri("https://localhost:7264/admin/logout")}
        };
        using var scope = app.Services.CreateScope();
        var applicationManager = scope.ServiceProvider.GetRequiredService<OpenIddictApplicationManager<Application>>();
        await applicationManager.CreateAsync(testApp);
        
        // Act / Assert
    }
}
```
I get this error
```
System.InvalidOperationException
The ConnectionString property has not been initialized.
   at Npgsql.ThrowHelper.ThrowInvalidOperationException(String message)
   at Npgsql.NpgsqlConnection.Open(Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlConnection.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnectionAsync(Boolean errorsExpected, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternalAsync(Boolean errorsExpected, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)
```

I would think (probably incorrectly) that since I'm running my AppHost project, that my DB and EntityFramework will be handled for me through appHost. When I run this project using aspire AppHost on my local machine, it works well and I can connect to my DB.
When I try to update the test to not explicitly add DataContext or AddOpenIddict in my test
```csharp
public class OidcClientProviderTest
{
    [Fact]
    public async Task ShouldLoginAndGetAccessToken()
    {
        // Arrange
        var appHost = await DistributedApplicationTestingBuilder.CreateAsync<Projects.AppHost>();

        await using var app = await appHost.BuildAsync();
        await app.StartAsync();

        // Test Application
        var testApp = new OpenIddictApplicationDescriptor
        {
            ClientId = "TestClient",
            ClientSecret = "TestClientSecret",
            Requirements = { OpenIddictConstants.Requirements.Features.ProofKeyForCodeExchange },
            DisplayName = "Administration",
            RedirectUris = {new Uri("https://localhost:7264/admin")},
            PostLogoutRedirectUris ={new Uri("https://localhost:7264/admin/logout")}
        };
        using var scope = app.Services.CreateScope();
        var applicationManager = scope.ServiceProvider.GetRequiredService<OpenIddictApplicationManager<Application>>();
        await applicationManager.CreateAsync(testApp);
    }
}
```
I get an error that suggests to add this.
```
System.InvalidOperationException
No service for type 'OpenIddict.Core.OpenIddictApplicationManager`1[Data.Security.Application]' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at IntegrationTests.OidcClientProviderTest.ShouldLoginAndGetAccessToken() in /Users/bates/Desktop/code/aspireProject/Services/Tests/IntegrationTests/OidcClientProviderTest.cs:line 120
   at IntegrationTests.OidcClientProviderTest.ShouldLoginAndGetAccessToken() in /Users/bates/Desktop/code/aspireProject/Services/Tests/IntegrationTests/OidcClientProviderTest.cs:line 121
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 253
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
```

The Actual Project I'm wanting to test is my Accounts Project, that is setup in this way, inside of my AppHost/Program.cs file
```csharp
var postgres = builder
    .AddPostgres(Resources.POSTGRES_SERVER, port: 15432)
    .WithDataVolume()
    .AddDatabase(Resources.POSTGRES_DATABASE, "FooBar");

var accounts = builder.AddProject<Accounts>(ServiceConstants.ACCOUNTS, HTTPS_PROFILE)
    .WithReference(redis)
    .WithReference(postgres)
    .WaitForCompletion(migrations); //Block this from running until migrations are finished.
```
This may not be an aspire thing, and more of a fundamental on dotnet misunderstanding, but I'm having a hard time extricating one from the other with these errors.

**Reply by davidfowl** [reply]
I'm trying to piece together that parts of this problem without the source. I think there might be a misunderstanding of what the `DistributedApplicationTestingBuilder` does. It's spinning up external processes that your tests can talk to. You cannot resolve services from your projects.

There's a pretty long thread here about integration testing https://github.com/dotnet/aspire/discussions/878. That might be what you are looking for.

**Reply by jhancock-taxa** [reply]
We would be fine if it just properly did ServiceDiscovery setup.

The use case is as follows:

Group of integration tests that depend on a specific set of state in a postgress database for real testing against all services and the state in the database.

Thus we need to:

1. Spin up the aspire host.
2. Create the data in the postgres database for the tests (i.e. creating seed data for the controlled tests)
3. Run the test using a Kiota client calling into services.

This tests:

1. The Kiota client is correct and points to the right spots and formats data correctly.
2. The auth client is connecting properly to our OpenIdConnect server, using password flow and getting a token.
3. The micro service can perform introspection on the user with credentials for the specific test.
4. The micro service responds accordingly with the correct known data from the seeded data and absolutely nothing else.

I would suggest that this is one of the primary expected use cases for this type of integration testing. It cannot be run without the orchestration and it means we must be able to get Service discovery working in this case to be able to do the seed data.

The only reason for the database in the integration tests per-say is to seed the data. But this is a pretty common scenario and even described in Microsoft Training Materials: https://learn.microsoft.com/en-us/ef/core/testing/testing-with-the-database (all be it without Aspire doing the service orchestration)

**Reply by jhancock-taxa** [reply]
Also, we'd like to be able to set the environment (aspnetcore_environment/netcore_environment) for the tests so that we can signal to the apphost that it's running in a test environment.

**Reply by davidfowl** [reply]
I don't really understand what the problems are given the exceptions on this thread, but if the goal is integration testing then I'd suggest looking at that thread to see what is possible today and if it satisfies your requirements:

https://github.com/dotnet/aspire/discussions/878#discussioncomment-9631749

**Reply by jhancock-taxa** [reply]
The exceptions are a red herring.  

The problem is:  DistributedApplicationTestingBuilder doesn't inject ServiceDiscovery properly:

With this in the apphost:

```
var pgBuilder = builder.AddPostgres("PostgresServer")
pgBuilder.AddDatabase("Postgres", "DBNameHere");

```

The following doesn't get any service discovery:
```
        var appHost =
            await DistributedApplicationTestingBuilder.CreateAsync<Projects.AppHost>();

        await using var aspireApp = await appHost.BuildAsync();
        await aspireApp.StartAsync();

```

At no point here with any of this, be it from appHost.Configuration or aspireApp.Services can you get any of the connection strings.

They're not anywhere that I can find in any DI Provider, trying to pull IOptions<ConnectionStrings> doesn't work, IConfiguration.GetSection("ConnectionStrings") is null, etc.

If DistributedApplicationTestingBuilder.CreateAsync properly did Service Discovery, then all of the rest of this ticket would be closed off.

We REALLY don't want to get to trying to create WebHost test stuff per that ticket.  We just want the aspire host running like normal and service discovery injected so that we can seed the database and get the project paths for httpclient.  The later works, but Postgres etc are not contributing connection strings.

(Ideally in test mode, all ports would be randomized and ensured to not be used yet of course)

**Comment by jhancock-taxa** [comment]
A little more context here:

The aspire app host after started doesn't have any of the connection strings in the Configuration at all. 

It does if you run it normally.

This is running from Rider with the Aspire plugin installed.  We've also tried running the tests from Vs code with the same results.  We're a mac shop so no ability to run in VS.net.

Docker desktop shows postgres running (and pg admin works fine, and shows that the migration job that we have run on apphost startup runs), so it appears thatDistributedApplicationTestingBuilder is not doing the connection string injection properly.

**Reply by DeveloperRegan** [reply]
I am also having the problem when I am running it on Visual Studio Community on Windows.

**Reply by v-bohatikov** [reply]
So, is there any progress with this? Because without being able to get underling connection strings, DistributedApplicationTestingBuilder have no sense at all. I'm currently trying to figure out how to get connection strings from azure simulator resources to create testing services, but I'm loosing in this battle.
